Describe how to run load tests and load test outcomes (#1312)
Azure/iotedge-lorawan-starterkit: docs/user-guide/testing/load_tests.md

## Running the load tests

In `LoRaWan.Tests.Simulation` you have access to a set of load tests that can be used to validate the performance of the system. If you have access, you can either run the load tests from the E2E CI by setting the `RunLoadTests` input parameter to `true`. Alternatively, you can run the tests on your local machine. If you decide to run the tests on your local machine, first to copy the `appsettings.json` from the Simulator project into a `appsettings.local.json` and make sure that this file is copied over to the output directory. Replace the tokens in the `appsettings.local.json`:

- To connect the load tests with a locally running LNS, replace the value in `LoadTestLnsEndpointsString` with something similar to `{"LoadTestLnsEndpointsString":"[\"ws://<lns-url>:5000\"]"}`. You can insert an arbitrary number of LNS stations into the array, and the load tests will connect stations to LNS using round-robin distribution.

> - The `DevicePrefix` can be any prefix that is used for the load test devices. If `CreateDevices` is set to `true`, it will create devices with this prefix in the IoT Hub that is referenced in your `appsettings.local.json`.

Reviewer: ```suggestion
- The `DevicePrefix` can be any prefix that is used for the test devices generating the load. If `CreateDevices` is set to `true`, it will create devices with this prefix in the IoT Hub that is referenced in your `appsettings.local.json`.
```

- Make sure that the `LeafDeviceGatewayID` matches the ID of one of the LNS that you connect to the load tests.

## Example load tests

All load tests use the `Connected_Factory_Load_Test_Scenario` load test scenario.

### January 25th 2022, load test of v2.0.0-beta1 in gateway mode (ENABLE_GATEWAY=true)

> We ran a set of load tests to ensure that we can support a certain amount of LoRa devices, depending on various parameters. In gateway mode, the edge hub [refreshes device twin in the background](https://github.com/Azure/iotedge/blob/master/doc/EnvironmentVariables.md#edgehub), which can result in a higher load on the IoT Hub and subsequent throttling operations due to IoT Hub quota. By taking into account these variables and KPI, we ran several load tests successfully and list the parameters we used below. For all tests we ran at a rate of 1 join request per second and 1 upstream message during the cache pre-warm phase (for more than 500 devices we use 4 requests per second during the warm-up phase to speed up the process), successively increasing the load to 10 messages per second. With this message rate, in general we saw the queue filling up too quickly on the Edge Hub. In such cases, we started to use lower message rates (3/second).

Reviewer: ```suggestion
We ran a set of load tests to ensure that we can support a certain amount of LoRa devices, depending on various parameters. In gateway mode, the edge hub [refreshes device twin in the background](https://github.com/Azure/iotedge/blob/master/doc/EnvironmentVariables.md#edgehub), which can result in a higher load on the IoT Hub and subsequent throttling operations due to IoT Hub quota. By taking into account these variables and KPI, we ran several load tests successfully with the parameters we used listed below. For all tests we ran at a rate of 1 join request per second and 1 upstream message during the cache pre-warm phase (for more than 500 devices we use 4 requests per second during the warm-up phase to speed up the process), successively increasing the load to 10 messages per second. With this message rate, in general we saw the queue filling up too quickly on the Edge Hub. In such cases, we started to use lower message rates (3/second).
```

Issues encountered in these load tests:

- [Unexpected ObjectDisposedException when IoT Hub is throttling · Issue #6042 · Azure/iotedge (github.com)](https://github.com/Azure/iotedge/issues/6042)
- When executing the load tests for more than 1000 devices, we started to see many `AmqpException` with the message `Unable to open Events link as the connection could not be authenticated`
- In case of two gateways and high message rates (10 messages per second), we saw that the delivery latency spiked for one of the two gateways. This behavior did not occur with lower message rates (3 messages per second). This correlated with the IoT Edge hub queue filling up, and as such it's likely that this caused the delivery latency spikes.

The memory/CPU usage of the IoT Edge host in gateway mode (for 900 devices) was fairly stable at around 1GB of memory used and an average CPU consumption of 40% on a Standard D2s v3 Debian 11 VM:

![gateway-mode-host-cpu-memory](..\..\images\lt-host-gateway-mode.png)

### January 21st 2022, load test of v2.0.0-beta1 in direct mode (ENABLE_GATEWAY=false)

> This load test used the default deduplication strategy. First, we deployed two LNS on Standard D2s v3 Debian 11 VMs (2vCPU, 8GB of memory). We simulated 1600 OTAA devices, distributed among 8 factories with 4 concentrators each. All devices are in reach of all concentrators within the same factory and send 5 messages each, starting with a join rate of 1.5 messages per second (to not hit the IoT Hub S1 quota) and then with progressively sending messages faster and faster (starting at 1.5 messages per second to pre-warm the cache and not hit IoT Hub S1 quota, and then successively increase the load to around 9.5 messages per second). Keep in mind that the effective message rate is higher than 9.5 messages per second, since every message is delivered to the gateways by the four concentrators per factory. The following bugs/issues became apparent in the load test:

Reviewer: ```suggestion
This load test used the default deduplication strategy. First, we deployed two LNS on Standard D2s v3 Debian 11 VMs (2vCPU, 8GB of memory). We simulated 1600 OTAA devices, distributed among 8 factories with 4 concentrators each. All devices are in reach of all concentrators within the same factory and send 5 messages each, starting with a join rate of 1.5 messages per second and then progressively sending messages faster and faster (starting at 1.5 messages per second to pre-warm the cache and not hit IoT Hub S1 quota, and then successively increase the load to around 9.5 messages per second). Keep in mind that the effective message rate is higher than 9.5 messages per second, since every message is delivered to the gateways by the four concentrators per factory. The following bugs/issues became apparent in the load test:
```

- [AMQP exceptions when running load tests · Issue #1348 · Azure/iotedge-lorawan-starterkit (github.com)](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1348)
- [AMQP exceptions when handling > 1000 devices · Issue #1337 · Azure/iotedge-lorawan-starterkit (github.com)](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1337)
- [We should send messages upstream when on DeduplicationStrategy Mark or None. · Issue #1032 · Azure/iotedge-lorawan-starterkit (github.com)](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1032)
- [Resubmit threshold does not consider deduplication strategy used · Issue #1334 · Azure/iotedge-lorawan-starterkit (github.com)](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1334)
- [Investigate memory evolution of LNS under load · Issue #1374 · Azure/iotedge-lorawan-starterkit (github.com)](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1374)

> In addition to the issues, for which we will not provider more details here, we discuss several performance/health indicators.

Reviewer: ```suggestion
In addition to these issues, for which we will not provide more details here, we discuss several performance/health indicators.
```

**Unhandled exceptions**: For a total of ca. 10k messages sent, all exceptions that we saw (ca. 48) were caused by one of the issues mentioned above.

> **D2CMessageDeliveryLatency**: The D2C message delivery latency took a distinct shape for the three phases: in the join phase, the average processing time was ca. 100ms, then for the first round of messages (cache pre-warming) the average delivery/response time was ca. 800ms on the gateway winning the race. As soon as the cache was warm, the response time dropped to ca. 450ms. Zero receive windows were missed for all ca. 10k messages.

Reviewer: ```suggestion
**D2CMessageDeliveryLatency**: The D2C message delivery latency took a distinct shape for the three phases: in the join phase, the average processing time was ca. 100ms, then for the first round of messages (cache pre-warming) the average delivery/response time was ca. 800ms on the gateway winning the race. As soon as the cache was warm, the response time dropped to ca. 450ms. **No receive windows were missed for all ca. 10k messages.**
```

![image-20220124131449660](..\..\images\lt-message-latency.png)

> **Memory and CPU usage**: CPU usage was fairly stable. However, while memory was staying between 100 and 130MB for the LNS during the entire load test. To ensure that we do not have a memory leak, we ran a longer load test over the course of several hours, during which memory was bounded at ca. 200MB and analysis of the Heap Dump revealed that the largest contributor to the Gen 2 Heap were the device connections (as expected), which the LNS manages in an internal cache.

Reviewer: ```suggestion
**Memory and CPU usage**: CPU usage was fairly stable, while memory was staying between 100 and 130MB for the LNS during the entire load test. To ensure that we do not have a memory leak, we ran a longer load test over the course of several hours, during which memory was bounded at ca. 200MB and analysis of the Heap Dump revealed that the largest contributor to the Gen 2 Heap were the device connections (as expected), which the LNS manages in an internal cache.
```
